[survey] Fix for JSONdata and LINST instruments

ridz1208 opened the pull request:
## Brief summary of changes
The survey instrument logic had fallen severely behind the latest standards. This PR adds support for LINST instrument pages, `getSubtestList()` which deprecated the `instrument_subtests` table in the database and JSON data saving and loading.

MAKE SURE TO REVIEW PR with "Hide whitespaces changes" as the indentation in files was fixed for readability

More details on changes:
1. fixes numbering issues on pages in survey instruments (causing next/prev page to not function properly)
2. fixes type hinting issues, more specifically on return values from `getPrevPage()` and `getNextPage()` functions
3. fixes saving/loading functionality to support JSON data in survey forms
4. cleans redundant page number variables and unused variables (ie. `$totalpages`)
5. fixes #7667 (hopefully)


#### Testing instructions (if applicable)

1. Follow survey testing plan
2. make sure to test for at least 1 of each of the following usecases
    - PHP instrument using SQL table for saving
    - LINST instrument using SQL table
    - PHP instrument using JSON saving in flag table
    - LINST instrument using JSON saving in flag table
    - single page instrument
    - multipage instrument

#### Link(s) to related issue(s)

* Resolves #7667

ridz1208 commented:
@driusan critical ?

ridz1208 commented on php/libraries/NDB_BVL_Instrument.class.inc:
```diff
@@ -1403,11 +1404,7 @@ abstract class NDB_BVL_Instrument extends NDB_Page
         $allValues = $this->getInstanceData();
-
-        if (!array_key_exists($field, $allValues)) {
-            throw new \OutOfBoundsException("Invalid field for instrument");
```
@driusan might be necessary for 23 as well

driusan commented:
@ridz1208 is going to split this up into multiple PRs. Parts of it (the NDB_BVL_Instrument part) are critical, parts of it (the survey part) can go into a bug fix release after being well tested, since the changes seem pretty significant.

ridz1208 commented:
needs rebase afte #7819 and #7820 are merged

ridz1208 commented:
blocked by #7818

ridz1208 commented:
@driusan I think this was deemed as not critical to release but it's still a valueable bugfix if we wanna assign it. It is rebased and ready to go

ridz1208 commented:
Hey @suzanne-lee given that you have the initial ticket for this. can you try out this fix ?

suzanne-lee commented:
For BMI Calculator, the top left corner now says "Page 1 of 4" rather than "Page 1 of 1". However, the save & continue button is no longer appearing and therefore cannot be fully tested. See screenshot:

<img width="1270" alt="Screen Shot 2022-01-11 at 9 37 21 AM" src="https://user-images.githubusercontent.com/29237731/148963492-4e4b96ec-ec69-46db-8631-be1f3b5d2a41.png">

ridz1208 commented:
@suzanne-lee should be ready again

ridz1208 commented:
@suzanne-lee I removed the print statement and fixed the page numbers shown.

As for your saving issue, I do not get that error when I test, here are a couple debugging steps you can take:
1. make sure your bmi.linst and bmi.meta files are freshly copied from the raisinbread directory
2. set the `jsonData` variable in `bmi.meta` to `true` and see if the pages save. (if they save then you might have an SQL issue in your bmi table)
3. set the `jsonData` back to `false` and see if the issue you get is related to the SQL table you have for the bmi
4. try other instruments that are non LINST


I tried all steps above and they all work on my VM. Also the error you reported seems to be linked to the getDataDictionary() function in the NDB_BVL_Instrument_LINST class. If all else fails, maybe you can try to print the return of that function and see why the Date_taken field is not in that array

suzanne-lee commented:
Still trying to figure out the Date_taken problem, but the page numbering has been resolved.

suzanne-lee commented:
I am able to view and submit all of the pages, as well as submit the entire form. However, the "Choose unit of measurement" question doesn't save; the element's name is supposed to be `unit_classification` but is `pageNum` on the browser (see below).

<img width="1291" alt="Screen Shot 2022-04-05 at 10 08 28 AM" src="https://user-images.githubusercontent.com/29237731/161773681-0bbd8b2a-ba64-4567-bc8f-98dec239d968.png">

cmadjar commented:
@ridz1208 there are conflicts with the base branch. Also, should there be an entry in the CHANGELOG for this bug fix?

ridz1208 commented:
@driusan conflicts resolved
